### DISCUSSION TITLE
Is nested filtering even possible with usual Prisma?

### DISCUSSION THREAD

**Discussion by BJladika** [prisma/prisma#7622] [Q&A] [answered]
Okay I have the next prisma.schema
```prisma
model umo_brspoints {
  id             Int            @id @default(autoincrement())
  points         Float
  checkpoint_id  Int
  course_id      Int
  student_id     Int            @unique
  umo_checkpoint umo_checkpoint @relation(fields: [checkpoint_id], references: [id])
  umo_course     umo_course     @relation(fields: [course_id], references: [id])
  umo_student    umo_student    @relation(fields: [student_id], references: [person_ptr_id])
}

model umo_checkpoint {
  id                  Int                   @id @default(autoincrement())
  name                String                @unique
  umo_brspoints       umo_brspoints[]
  umo_coursemaxpoints umo_coursemaxpoints[]
}

model umo_course {
  id                    Int                   @id @default(autoincrement())
  discipline_detail_id  Int
  group_id              Int
  umo_disciplinedetails umo_disciplinedetails @relation(fields: [discipline_detail_id], references: [id])
  umo_group             umo_group             @relation(fields: [group_id], references: [id])
  umo_brspoints         umo_brspoints[]
}

model umo_discipline {
  id                    Int                     @id @default(autoincrement())
  Name                  String
  umo_disciplinedetails umo_disciplinedetails[]
}

model umo_disciplinedetails {
  id             Int            @id @default(autoincrement())
  discipline_id  Int
  semester_id    Int?
  umo_discipline umo_discipline @relation(fields: [discipline_id], references: [id])
  umo_semester   umo_semester?  @relation(fields: [semester_id], references: [id])
  umo_course     umo_course[]
}

model umo_group {
  id             Int             @id @default(autoincrement())
  Name           String
  umo_course     umo_course[]
  umo_grouplist  umo_grouplist[]
}

model umo_grouplist {
  id          Int                 @id @default(autoincrement())
  group_id    Int
  student_id  Int
  umo_group   umo_group           @relation(fields: [group_id], references: [id])
  umo_student umo_student         @relation(fields: [student_id], references: [person_ptr_id])
}

model umo_person {
  id          Int          @id @default(autoincrement())
  FIO         String
  last_name   String
  first_name  String
  second_name String
  maiden_name String
  umo_student umo_student?
  umo_teacher umo_teacher?
}

model umo_semester {
  id                    Int                     @id @default(autoincrement())
  name                  String                  @unique
  umo_disciplinedetails umo_disciplinedetails[]
}

model umo_student {
  person_ptr_id Int             @id @default(autoincrement())
  student_id    String
  umo_person    umo_person      @relation(fields: [person_ptr_id], references: [id])
  umo_brspoints umo_brspoints[]
  umo_grouplist umo_grouplist[]
}
```
Its scheme shortly would like looks like this (model was a bit simplified to contain less data for better readability)
![Umo_Schema](https://user-images.githubusercontent.com/23732658/121724688-2e89aa80-cb23-11eb-8650-fe10cac71528.png)

So in short I want to make a query with filtering umo_semester.name and umo_checkpoint.name but start from the group, query would look like this
```graphql
query{
  umo_group(id:X,umo_semester.name:"Y",umo_checkpoint.name:"Z"){
    umo_grouplist{
      umo_student{
        umo_person{
          FIO
        }
        umo_brspoints{
          points
          umo_checkpoint
        }
        umo_course{
          umo_discipline_details{
            umo_discipline{
              Name
            }
            umo_semester{
              name
            }
          }
        }
      }
    }
  }
}
```
smth like that (maybe without semester name and checkpoint name), in SQL it'd look like smth like this
```sql
Select P.FIO, D.Name, SEM.name, BS.points,  FROM
umo_group as G,
umo_grouplist as GL,
umo_student as S,
umo_person as P,
umo_brspoints as BS,
umo_course as C,
umo_disciplinedetails as DD,
umo_discipline as D,
umo_semester as SEM,
umo_checkpoint as CHK
WHERE
SEM.name = "Z"
AND
CHK.name = "Y"
AND
G.id = X
AND
G.id = GL.group_id
AND
GL.student_id = S.person_ptr_id
AND
S.person_ptr_id = P.id
AND
S.person_ptr_id = BS.student_id
AND
BS.course_id = C.id
AND
C.discipline_detail_id = DD.id
AND
DD.semester_id = SEM.id
AND
DD.discipline_id = D.id
AND
CHK.id = BS.checkpoint_id
```
So is it achievable with prisma? or maybe make these filters from umo_person and filter it with group_name without actually calling group name?

**Comment by BJladika** [comment, marked as answer]
I tried to make resolver smth like that (returns persons on group_name, semester name and checkpoint name selected like)
with type query ` umo_personsUnique(group_name: String, semester_name: String, checkpoint_name: String): [Person]!`
```javascript
    umo_personsUnique: (_parent, _args, context) => {
      return context.prisma2.umo_person.findMany({
        where:{
          umo_student:{
            every:{
              umo_grouplist:{
                umo_group:{
                  Name:{
                    contains: _args.group_name
                  }
                }
              }
            },
            every:{
              umo_brspoints:{
                umo_checkpoint:{
                  name:{
                    contains: _args.checkpoint_name
                  }
                },
                umo_course:{
                  umo_disciplinedetails:{
                    umo_semester:{
                      name:{
                        contains: _args.semester_name
                      }
                    }
                  }
                }
              }
            }
          }
        }
      })
    },
```
But I'm receiving error ```"message": "\nInvalid `prisma.umo_person.findMany()` invocation:\n\n{\n  where: {\n    umo_student: {\n      every: {\n      ~~~~~\n        umo_brspoints: {\n          umo_checkpoint: {\n            name: {\n              contains: 'Рубежный срез'\n            }\n          },\n          umo_course: {\n            umo_disciplinedetails: {\n              umo_semester: {\n                name: {\n                  contains: '7'\n                }\n              }\n            }\n          }\n        }\n      }\n    }\n  }\n}\n\nUnknown arg `every` in where.umo_student.every for type Umo_studentRelationFilter. Did you mean `select`? Available args:\ntype Umo_studentRelationFilter {\n  is?: umo_studentWhereInput\n  isNot?: umo_studentWhereInput\n}\n\n",```
Am I doing smth wrong?

**Reply by BJladika** [reply]
Okay, after some thoughts I understood that it was too overcomplicated, like why I even need to go through such tree.

So I decided to make umo_person filtering via group like (and then make some subfiltering with deeper resolvers)
```
query{
 umo_person(group_id: X){
  FIO
}
```

And made resolver
```javascript
    umo_personsUnique: (_parent, _args, context) => {
      return context.prisma2.umo_person.findMany({
        where:{
          umo_student:{
            umo_grouplist:{
              every:{
                umo_group:{
                  id:{
                    contains: _args.group_id
                  }
                }
              }
            }
          },
        }
      })
    },
```
but I'm receiving such an error and I dont understand what I made it wrong, any ideas?
```"message": "\nInvalid `prisma.umo_person.findMany()` invocation:\n\n{\n  where: {\n    umo_student: {\n      umo_grouplist: {\n      ~~~~~~~~~~~~~\n        every: {\n          umo_group: {\n            id: {\n              contains: 4142\n            }\n          }\n        }\n      }\n    }\n  }\n}\n\nUnknown arg `umo_grouplist` in where.umo_student.umo_grouplist for type Umo_studentRelationFilter. Available args:\n\ntype Umo_studentRelationFilter {\n  is?: umo_studentWhereInput\n  isNot?: umo_studentWhereInput\n}\n\n"```

**Reply by BJladika** [reply]
Ah, after some analysis I've understood that using operator `contains` for id is wrong thing.
Either it must be
```
id: _args.group_id
```
or 
```
id: {
 equals: _args.group_id
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
